File: Chapter 4 - Trees and Graphs/List Of Depths/list_of_depths.py
from collections import deque
class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None

#for deque use (append && popleft) or (appendleft && pop)
#naive approach to illustrate Linked Lists on each level
def levelOrder(Node):
    if Node is None:
        return
    queue, linkedlist = deque(), deque()
    queue.append(Node)
    queue.append(None)
    level = 1
    while queue:
        x = queue.popleft()
        if x == None:
            if queue:
                queue.append(None)
            while len(linkedlist) != 0:
                if len(linkedlist) == 1:
                    print(linkedlist.popleft().data)
                else:
                    print(linkedlist.popleft().data, end = '->')
        else:
            linkedlist.append(x)
            if x.left is not None:
                queue.append(x.left)
            if x.right is not None:
                queue.append(x.right)

if __name__ == "__main__":
    r = TreeNode(1)
    r.left = TreeNode(2)
    r.right = TreeNode(3)
    r.left.left = TreeNode(4)
    r.left.right = TreeNode(5)
    r.right.left = TreeNode(6)
    r.right.right = TreeNode(7)
    levelOrder(r)

File: Chapter 4 - Trees and Graphs/List Of Depths/list_of_depths.cpp
//https://ide.geeksforgeeks.org/BvwxQrFKmN
#include<bits/stdc++.h>
using namespace std;
struct Node{
    int data;
    Node *left, *right;
};

Node *newNode(int data){
    Node *curr = new Node();
    curr->data = data;
    curr->left = curr->right = NULL;
    return curr;
}

//Non-recursive
void levelOrder(Node *root){
    if(root == NULL)return;
    queue<Node *> q;
    vector<Node *> level;
    q.push(root);
    q.push(NULL);
    int height = 1;
    while(!q.empty()){
        Node *curr = q.front();
        q.pop();
        if(curr == NULL){
            if(!q.empty()){
                q.push(NULL);
                height++;
            }
            for(auto x=0; x<level.size(); x++)
                if(x == level.size() - 1)
                    cout<<level[x]->data<<"\n";
                else
                    cout<<level[x]->data<<"->";
            level.clear();
        }
        else{
            level.push_back(curr);
            if(curr->left) 
                q.push(curr->left);
            if(curr->right) 
                q.push(curr->right);
        }
    }
}

int main(){
    Node *root = newNode(1);
    root->left = newNode(2);
    root->right = newNode(3);
    root->left->left = newNode(4);
    root->left->right = newNode(5);
    root->right->left = newNode(6);
    root->right->right = newNode(7);
    levelOrder(root);
}